integration/spark/src/main/scala/org/apache/spark/util/PartitionCacheManager.scala

```diff
@@ -0,0 +1,181 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+import java.net.URI
+import java.util
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.log4j.Logger
+import org.apache.spark.sql.catalyst.catalog.{CatalogStorageFormat, CatalogTablePartition}
+
+import org.apache.carbondata.common.logging.LogServiceFactory
+import org.apache.carbondata.core.cache.{Cache, Cacheable, CarbonLRUCache}
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.datastore.impl.FileFactory
+import org.apache.carbondata.core.index.Segment
+import org.apache.carbondata.core.metadata.{AbsoluteTableIdentifier, SegmentFileStore}
+import org.apache.carbondata.core.statusmanager.{LoadMetadataDetails, SegmentStatusManager}
+import org.apache.carbondata.core.util.path.CarbonTablePath
+
+object PartitionCacheManager extends Cache[PartitionCacheKey,
+  java.util.List[CatalogTablePartition]] {
+
+  private val CACHE = new CarbonLRUCache(
+    CarbonCommonConstants.CARBON_PARTITION_MAX_DRIVER_LRU_CACHE_SIZE,
+    CarbonCommonConstants.CARBON_MAX_LRU_CACHE_SIZE_DEFAULT)
+
+  val LOGGER: Logger = LogServiceFactory.getLogService(this.getClass.getName)
+
+  def get(identifier: PartitionCacheKey): java.util.List[CatalogTablePartition] = {
+    LOGGER.info("Reading partition values from store")
+    // read the tableStatus file to get valid and invalid segments
+    val validInvalidSegments = new SegmentStatusManager(AbsoluteTableIdentifier.from(
+      identifier.tablePath, null, null, identifier.tableId))
+      .getValidAndInvalidSegments
+    val cacheablePartitionSpecs = validInvalidSegments.getValidSegments.asScala.map { segment =>
+      val segmentFileName = segment.getSegmentFileName
+      val segmentFilePath = FileFactory.getCarbonFile(
+        CarbonTablePath.getSegmentFilePath(identifier.tablePath, segmentFileName))
+      // read the last modified time
+      val segmentFileModifiedTime = segmentFilePath.getLastModifiedTime
+      val existingCache = CACHE.get(identifier.tableId)
+      if (existingCache != null) {
+        val segmentCache = CACHE.get(identifier.tableId).asInstanceOf[CacheablePartitionSpec]
+          .partitionSpecs.get(segment.getSegmentNo)
+        segmentCache match {
+          case Some(c) =>
+            // check if cache needs to be updated
+            if (segmentFileModifiedTime > c._2) {
+              (segment.getSegmentNo, (readPartition(identifier,
+                segmentFilePath.getAbsolutePath), segmentFileModifiedTime))
+            } else {
+              (segment.getSegmentNo, c)
+            }
+          case None =>
+            (segment.getSegmentNo, (readPartition(identifier,
+              segmentFilePath.getAbsolutePath), segmentFileModifiedTime))
+        }
+      } else {
+        // read the partitions if not available in cache.
+        (segment.getSegmentNo, (readPartition(identifier,
```

Comment:
readPartition reads the segment file. 
so basically first time we read all the segment files of the table.

I have seen the users with 200K segments in one table stored in cloud. query takes more than 5 hours for them if we do this.

So, need to change the key as partition spec and cache only matched partitions

@kumarvishal09 , @QiangCai @marchpure : what is your opinion on this ?

Reply:
agree with you. 
also avoid to use getAbsolutePath function，which has high IO overhead

Reply:
As you mentioned first time it will take lot of time, but query time we will get benefit. You said matched partitions and since here for every load we are trying to cache, we wont be able to get any matched partitions right?

It is something similar to our Indexes(MV, SI) etc where we load all segment at once in first time and from next time we do incremental, as @QiangCai mentioned. So here from subsequent loads, only that segment should be considered, which i think its being taken care.

@ajantha-bhat are you trying to say to change to handle in query or load itself? Correct me, if i'm wrong.

Reply:
and i also checked now, this caching happens in flow query too, so you can consider my reply specific to load.

Reply:
@akashrn5 : yeah, my problem scenario is first time query or upgrade, taking huge time (near to 5 hours for first time query) is still not ok.  In this scenario, I feel reading from hive metastore itself is very fast compared to number of IO from the cloud storage

Reply:
@ajantha-bhat This is disabled by default, so if the user feels that hive metastore getPartitions is slow, can enable this to improve performance.

Reply:
ok. If the user feels hive metastore getPartitions is slow, still he cannot enable this in cloud scenario if he had the huge segments as it will be slower than hive metastore. 

As the feature is disabled by default. Ok for me.

Reply:
@ajantha-bhat

Reply:
@ajantha-bhat 

`sparkSession.sessionState.catalog.listPartitionsByFilter(identifier, partitionFilters)` is already being used for to filter the partitions based on the filter, but in concurrent queries as apark has to take a lock on metastore before accessing it, therefore the query performance was being degraded.

And no user will have 200k segment, they need to anyways compact otherwise the query performance would be degraded

Reply:
we have users with 200k segments. But the data is partitioned by date and time. so all queries will have partition columns filter. so they query-specific segments every time (without any degrade in query time)

Reply:
Yeah, because they are using presto, where compaction is not yet supported.

If they are using hivemetastore surely they will face degrade in concurrent queries